### PR TITLE
[WIP] made whole adaptive workflow jax/jit

### PR DESCRIPTION
**Pull request by mrshirts**
Converted the entire loop in adaptive to jax/jit. 
See discussion on issue #340 for discussion of timings.

Intended for inspection, not merging for now. 

Failing because jax/jit isn't loaded via conda (not clear it can be right now, conda-forge install on my machine didn't work, had to pip install).

### CODE REVIEW

File: pymbar/mbar_solvers.py [hunk 1]
```diff
@@ -3,7 +3,12 @@
 import numpy as np
 import math
 import scipy.optimize
-from pymbar.utils import ensure_type, logsumexp, check_w_normalized
+from pymbar.utils import ensure_type, check_w_normalized
+import jax
+from jax.scipy.special import logsumexp
+from jax.ops import index_update, index
+from jax.config import config; config.update("jax_enable_x64", True)
```
Comment (proteneer) [inline pymbar/mbar_solvers.py:10]: This will slow things down dramatically, esp. on the GPU. Do you really need 64bit precision?

File: pymbar/mbar_solvers.py [hunk 2]
```diff
@@ -289,105 +297,87 @@ def adaptive(u_kn, N_k, f_k, tol=1.0e-12, options=None):
...
+    results['f_k'] = f_k
+    results['max_delta'] = max_delta
+    results['iterations'] = iteration
+    results['gnorms_sci'] = gnorms_sci
+    results['gnorms_nr'] = gnorms_nr
+    results['which_method'] = which_method
+
+    return results
 
+jit_core_adaptive = jax.jit(core_adaptive,static_argnums=(0,1,2,3,4,5,))
```
Comment (proteneer) [inline pymbar/mbar_solvers.py:373]: When you call jax.jit, the first invocation will be exceedingly slow it's compiling the JIT kernels. So if you're benchmarking you should call this *more* than once to amortize out the jit time.

Futhermore, if the arguments have shapes that contain (None,) in one of the shapes, a recompilation is required due to an XLA requirement for each specialization of (None,) into a known shape.

File: pymbar/mbar_solvers.py [hunk 3]
```diff
@@ -289,105 +297,87 @@ def adaptive(u_kn, N_k, f_k, tol=1.0e-12, options=None):
...
+    nr_iter = 0
+    sci_iter = 0
 
-    for iteration in range(0, options["maximum_iterations"]):
+    which_method = []
+    gnorms_sci = []
+    gnorms_nr = []
 
+    # Perform Newton-Raphson iterations (with sci computed on the way)
+    for iteration in range(0, maximum_iterations):
```
Comment (proteneer) [inline pymbar/mbar_solvers.py:320]: can you write this as jax.lax.while_loop? Just jit-compiling this for loop while be a nightmare for XLA

File: pymbar/mbar_solvers.py (hunk 1 above)
Comment (mrshirts) [reply pymbar/mbar_solvers.py:10]: Probably yes - the algorithm doesn't seem to converge well with floats.  I can poke around and see what the deal is.

File: pymbar/mbar_solvers.py [hunk 4]
```diff
@@ -145,16 +149,28 @@ def mbar_objective_and_gradient(u_kn, N_k, f_k):
...
-    grad = -1 * N_k * (1.0 - np.exp(f_k + log_numerator_k))
+    grad = -1 * N_k * (1.0 - jnp.exp(f_k + log_numerator_k))
 
     obj = math.fsum(log_denominator_n) - N_k.dot(f_k)
 
     return obj, grad
 
+def jax_mbar_hessian(u_kn, N_k, f_k):
+
+    jNk = 1.0*N_k
```
Comment (proteneer) [inline pymbar/mbar_solvers.py:163]: what's the point of 1.0*?

File: pymbar/mbar_solvers.py (hunk 2 above)
Comment (mrshirts) [reply pymbar/mbar_solvers.py:373]: Ah, this explains why the partial jitification works better- then it's calling different functions multiple times.  Though only 7-8 times in many cases.

Comment (mrshirts) [reply pymbar/mbar_solvers.py:373]: I will test running mbar initialization several times with different data sets in the same script to see if that helps.

File: pymbar/mbar_solvers.py (hunk 3 above)
Comment (mrshirts) [reply pymbar/mbar_solvers.py:320]: > can you write this as jax.lax.while_loop?

I can take a look, though the fact that it's on the adaptive loop that is generally only called once in pymbar means it won't be as useful to do this.

File: pymbar/mbar_solvers.py (hunk 2 above)
Comment (proteneer) [reply pymbar/mbar_solvers.py:373]: I also noticed you're declaring every arg to be static, which means that unless you're passing in the exact args you'll be triggering a recompilation. Why did you have to declare them to be static to start with?

Comment (proteneer) [reply pymbar/mbar_solvers.py:373]: U_kn probably shouldn't be static if you're trying to pass in results from different runs?

File: pymbar/mbar_solvers.py (hunk 4 above)
Comment (mrshirts) [reply pymbar/mbar_solvers.py:163]: N_k is an int, jit (or maybe Jax, I can't recall) complains if it's not converted to float first.  I wasn't able to find the best direct conversion to float function that made jit happy. so I just multiplied by 1.0 to do it automatically.

File: pymbar/mbar_solvers.py (hunk 3 above)
Comment (proteneer) [reply pymbar/mbar_solvers.py:320]: maybe it makes sense then to just jit the body of this function? (i.e. move everything inside the for loop to a separate function and just jit that)

Comment (mrshirts) [reply pymbar/mbar_solvers.py:320]: Ah, yes, that makes a lot of sense given the constraints. I'll try that.

File: pymbar/mbar_solvers.py (hunk 2 above)
Comment (mrshirts) [reply pymbar/mbar_solvers.py:373]: I had to declare it static because otherwise, gnorm_sci and gnorm_nr end up as traced arrays instead of concrete arrays (since they are functions of traced arrays), and jit refuses to compile the conditional comparing the two, which decides the branch to take, unless they are declared static.

Comment (mrshirts) [reply pymbar/mbar_solvers.py:373]: @proteneer had a good suggestion for handling the conditional:

import jax
import jax.numpy as jnp
# can't be JIT'd
def foo(a, b, cond_a, cond_b):
    if cond_a < cond_b:
        return a+b
    else:
        return a-b
# foo_jit = jax.jit(foo) # fails
print(foo(0., 1., 2., 3.))
# can be JIT'd
def bar(a, b, cond_a, cond_b):
    return jnp.where(cond_a < cond_b, a+b, a-b)
bar_jit = jax.jit(bar)
print(bar_jit(0., 1., 2., 3.))

File: pymbar/mbar_solvers.py (hunk 1 above)
Comment (mrshirts) [reply pymbar/mbar_solvers.py:10]: Still need to work on this . . .

File: pymbar/mbar_solvers.py (hunk 2 above)
Comment (mrshirts) [reply pymbar/mbar_solvers.py:373]: So, if we accelerate just the inner loop, then we can pull the conditional out of the accelerated code with little loss of timing - all of the conditionals are just comparing floats and assigning values, so there is little use in accelerating them.

**Comment by mrshirts** [comment]
Closing in favor of one that just accelerates the inner loop.
